src/Test/L0/Worker/TaskManagerL0.cs

```diff
@@ -131,5 +132,153 @@ private Stream GetZipStream()
+                string bingDestPath = taskManager.GetDestinationPath(bingTask.Name, bingTask.Id, bingVersion);
+                string pingDestPath = taskManager.GetDestinationPath(pingTask.Name, pingTask.Id, pingVersion);
+                Assert.True(!File.Exists(Path.Combine(bingDestPath, "task.json")));
+                Assert.True(!File.Exists(Path.Combine(pingDestPath, "task.json")));
+                //assert download was invoked only once, because the first task cancelled the second task download
+                _jobServer
+                    .Verify(x => x.GetTaskContentZipAsync(It.IsAny<Guid>(), It.IsAny<TaskVersion>(), hc.CancellationToken), Times.Once());
+
+                //Cleanup
+                string zipFile = GetZipFolderName() + ".zip";
+                if (File.Exists(zipFile))
+                {
+                    File.Delete(zipFile);
+                }
+            }
+        }
+
+        //Test how exceptions are propagated to the caller.
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public async void TestNetworkErrorWhileDownloadingTask()
+        {
+            //Arrange
+            using (var hc = new TestHostContext(this))
+            {
+                hc.SetSingleton<IJobServer>(_jobServer.Object);
+                hc.SetSingleton<IConfigurationStore>(_configurationStore.Object);
+                var taskManager = new TaskManager();
+                taskManager.Initialize(hc);
+
+                _configurationStore.Setup(x => x.GetSettings()).Returns(
+                 new AgentSettings
+                 {
+                     WorkFolder = WorkFolderName
+                 });
+
+                string workingFolder = IOUtil.GetWorkPath(hc);
+                if (Directory.Exists(workingFolder))
+                {
+                    Directory.Delete(IOUtil.GetWorkPath(hc), true);
+                }
+
+                var pingTask = new TaskInstance()
+                {
+                    Enabled = true,
+                    Name = "Ping",
+                    Version = "0.1.1",
+                    Id = Guid.NewGuid()
+                };
+
+                var pingVersion = new TaskVersion(pingTask.Version);
+
+                _jobServer
+                    .Setup(x => x.GetTaskContentZipAsync(It.IsAny<Guid>(), It.IsAny<TaskVersion>(), hc.CancellationToken))
+                    .Returns((Guid taskId, TaskVersion taskVersion, CancellationToken token) =>
+                    {
+                        throw new System.Net.Http.HttpRequestException("simulated network error");
+                    });
+
+                var tasks = new List<TaskInstance>(new TaskInstance[] { pingTask });
+
+                //Act
+                //EnsureTasksExist should throw a HttpRequestException exception
+                Task ensureExistTask = taskManager.EnsureTasksExist(tasks);
+                Task[] taskToWait = { ensureExistTask, Task.Delay(2000) };
```

Comment:
This is to avoid test blocking forever in case of error. we will talk offline to and may change the pattern to more readable that you suggest or not do it at all if it is not a big concern.